qkernel/src/interrupt/aarch64/fault.rs

```diff
@@ -33,115 +33,81 @@ use crate::qlib::kernel::arch::__arch::mm::pagetable::{PageTableEntry, PageTable
 
 use crate::{MainRun, SHARESPACE, panic};
 
-#[repr(u64)]
-enum PageFaultErrorFlags {
-    FaultPermission  = 1 << 0,
-    FaultWrite       = 1 << 1,
-    FaultUserMode    = 1 << 2,
-    FaultAddressSize = 1 << 3,
-    FaultInstruction = 1 << 4,
-    FaultTranslation = 1 << 5,
-    FaultAccessFlag  = 1 << 6,
-    FaultKillItFlag  = 1 << 7, //Unhandled, Unforeseen, Unforgiven! Kill it.
+bitflags! {
+    #[repr(transparent)]
+    pub struct PageFaultErrorCode: u64 {
+        // Types (should be mutual exclusive)
+        const NO_HANDLER            = 1 << 0; //Unhandled, Unforeseen, Unforgiven! Kill it.
+        const PROTECTION_VIOLATION  = 1 << 1;
+        const ADDRESS_SIZE          = 1 << 2;
+        const TRANSLATION           = 1 << 3;
+        const ACCESS_FLAG           = 1 << 4;
+        const TYPE_MASK             = 0xff;
+        // Attributes
+        const CAUSED_BY_WRITE       = 1 << 8;
+        const USER_MODE             = 1 << 9;
+        const INSTRUCTION_FETCH     = 1 << 10;
+    }
 }
 
-#[derive(Debug)]
-pub struct PageFaultErrorCode (u64);
-
-impl LowerHex for PageFaultErrorCode {
-    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
-        write!(f, "{:#x}", self.0)
-    }
+impl EsrDefs {
+    // ESR.ISS.xxSC[5:2], Data/Instruction Fault Status Code WITHOUT translation level.
+    pub const SC_MASK: u64 = 0x3c;
+    pub const SC_ADDRESS_SIZE_FAULT: u64 = 0x0;
+    pub const SC_TRANSLATION_FAULT:  u64 = 0x4;
+    pub const SC_ACCESS_FLAG_FAULT:  u64 = 0x8;
+    pub const SC_PERMISSION_FAULT:   u64 = 0xc;
 }
 
 impl PageFaultErrorCode {
-    //
-    // NOTE: ISS[5:0] - Reveal type not level.
-    // NOTE: Maybe not correct place to have them,
-    //       still better than magic numbers.
-    //
-    #[allow(non_upper_case_globals)]
-    pub const GEN_xxSC_MASK: u64 = 0x3c; // Ignore the ll-bits
-    pub const GEN_ADDRESS_SIZE_FAULT: u64 = 0x0;
-    pub const GEN_TRANSLATION_FAULT:  u64 = 0x4;
-    pub const GEN_ACCESS_FLAG_FAULT:  u64 = 0x8;
-    pub const GEN_PERMISSION_FAULT:   u64 = 0xc;
-
+    pub fn get_type(&self) -> Self{
+        *self & Self::TYPE_MASK
+    }
 
-    pub fn new(from_user: bool, esr: u64) -> Self {
-        let mut fault_flags = PageFaultErrorCode(0);
+    pub fn new(from_user: bool, is_instr: bool, esr: u64) -> Self {
```

Comment:
we already know if that's an instruction abort along the way. There is no need to calculate it again from the ESR, which is pretty nasty code.